src/Internal/Payment/State/StateFactory.php

```diff
@@ -0,0 +1,79 @@
+<?php
+/**
+ * Class StateFactory
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment\State;
+
+use WCPay\Container;
+use WCPay\Vendor\League\Container\Exception\ContainerException;
+use WCPay\Internal\Payment\Exception\StateTransitionException;
+use WCPay\Internal\Payment\PaymentContext;
+
+/**
+ * A factory for payment states.
+ *
+ * The main purpose of this class is to work as an extension of the
+ * WooPayments dependency container, only allowing access to payment states
+ */
+class StateFactory {
+	/**
+	 * Holds the WooPayments DI container.
+	 *
+	 * @var Container
+	 */
+	private $container;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param Container $container Dependency container.
+	 */
+	public function __construct( Container $container ) {
+		$this->container = $container;
+	}
+
+	/**
+	 * Creates a new state based on class name.
+	 *
+	 * @param string         $state_class Name of the state class.
+	 * @param PaymentContext $context     Context for the new state.
+	 * @return PaymentState               The generated payment state instance.
+	 * @throws StateTransitionException   In case the class was not found or invalid.
+	 */
+	public function create_state( string $state_class, PaymentContext $context ): PaymentState {
+		/**
+		 * If resolved, this will always be a payment state.
+		 *
+		 * @var PaymentState|null
+		 */
+		$state = null;
+
+		if ( is_subclass_of( $state_class, PaymentState::class ) ) {
+			try {
+				// Load the state from the container.
+				$state = $this->container->get( $state_class );
+			} catch ( ContainerException $e ) {
+				// A proper exception would be thrown if the state is null.
+				$state = null;
+			}
+		}
+
+		if ( is_null( $state ) ) {
+			throw new StateTransitionException(
+				sprintf(
+					// Translators: %1$s is the PHP class for a new payment state, %1$s is the state base class.
+					__( 'The class %1$s either does not exist, or is not a subclass of %2$s', 'woocommerce-payments' ),
+					$state_class,
+					PaymentState::class
+				)
+			);
+		}
+
+		$state->set_context( $context );
+
+		return $state;
```

Comment:
Done in https://github.com/Automattic/woocommerce-payments/pull/7318/commits/56933778b2c3bf6cfb97fad2e4e5204e4a74fd07. The nesting was there to avoid multiple exceptions within such a simple method.

FWIW, I'm using the same exception class (`StateTransitionException` now) even for container issues. Ideally, this could bubble up the container exception, but as you said, it should not happen often. Moreover, we'd need to add an additional `@throws` line to each transition method of the state class, and we've already got long doc blocks there.

Reply:
Letting the container exception bubble up seems like the way to go. Because of the Psalm rule, it will require a `@throws` tag for every transition method, but that should be fine. Will do tomorrow.

Reply:
Done: https://github.com/Automattic/woocommerce-payments/pull/7318/commits/abc81746f8f9bfd7cf57f398af707204242e6c0d